src/Shared/FileMatcher.cs

```diff
@@ -764,6 +766,11 @@ bool needsRecursion
             /// </summary>
             public string Filespec { get; }
             /// <summary>
+            /// Holds the directory pattern for globs like **/{pattern}/**, i.e. when we're looking for a matching directory name
```

Comment:
What if there's multiple directory patterns like `**/{pattern1}/**/{pattern2}/**`

Reply:
Oh, as directory traversal descends one FilesSearchData instance will have it set to `pattern1`, then a later instance will have it set to `pattern2`, right? Remembering this code brings up some PTSD :P

Reply:
I am not optimizing `**/{pattern1}/**/{pattern2}/**` in this PR. This one I'm fine being handled by the slow regex path because it's not used by the .NET SDK and overall doesn't feel like something common (I have no data to back it up, just a gut feeling).

I think that you're right that it could be done without regex'es, although I'm sure there are many gotchas that prevent it from working simply by peeling the path components one by one as we're descending. For example, the `**` should not do eager matching so you still generally need a "lookahead". It may happen that you've descended multiple levels and everything matched but then you find a mismatch that could have been prevented by going back and changing your mind about a component matching against `**` or not.

Reply:
Might help to make the comment more explicit that the field is only applicable for that particular optimization and it's not part of the general flow

Reply:
Done, thank you.